src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+
+    @staticmethod
+    def _walk_path(n, path):
+        for attr, idx in path:
+            n = getattr(n, attr)
+            if idx is not None:
+                n = n[idx]
+        return n
+
+    def _rewrite_node(self, fn):
+        """
+        Applies `fn` to the AST node containing the cursor. The callback gets
+        the following arguments:
+        1. The parent node
+        2. The relevant list of children
+        3. The last step in the path (the attribute and index info)
+        The callback is expected to return a single, updated node to be placed
+        in the new tree.
+        """
+        assert isinstance(self, (Node, Block, Gap))
+
+        def impl(node, path):
+            if len(path) == 1:
+                return fn(node)
+
+            (attr, i), path = path[0], path[1:]
+            children = getattr(node, attr)
+
+            if i is None:
+                return node.update(**{attr: impl(children, path)})
+
+            return node.update(**{
+                attr: children[:i] + [impl(children[i], path)] + children[i + 1:]
+            })
+
+        return impl(self.proc().INTERNAL_proc(), self._path)
+
+    def _make_forward(self, new_proc, fwd_node, fwd_gap, fwd_sel):
+        orig_proc = self._proc
+        depth = len(self._path)
+        attr = self._path[depth - 1][0]
+
+        def forward(cursor: Cursor) -> Cursor:
+            if cursor._proc != orig_proc:
+                raise InvalidCursorError('cannot forward unknown procs')
+
+            # TODO: use attrs + attrs.evolve
+            def evolve(p):
+                return type(cursor)(new_proc, p)
+
+            old_path = cursor._path
+
+            if len(old_path) < depth:
+                # Too shallow
+                return evolve(old_path)
+
+            old_attr, old_idx = old_path[depth - 1]
+
+            if old_attr != attr:
+                # At least as deep, but wrong branch
+                return evolve(old_path)
+
+            if len(old_path) > depth or isinstance(cursor, Node):
+                idx = fwd_node(old_idx)
+            elif isinstance(cursor, Gap):
+                idx = fwd_gap(old_idx)
+            else:
+                assert isinstance(cursor, Block)
+                idx = fwd_sel(old_idx)
+
+            return evolve(old_path[:depth - 1] + [(attr, idx)] + old_path[depth:])
+
+        return forward
+
+
+@dataclass
+class Block(Cursor):
+    _path: list[tuple[str, Union[int, range]]]  # is range iff last entry
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        attr, _range = self._path[-1]
+        assert len(_range) > 0
+        return self.parent().child(attr, _range.start).before(dist)
```

Comment:
We should probably implement these without the `parent().child(...)` pattern?
e.g. can't we just directly update the path to construct the `start` or `stop` node in the block, and then call `before` or `after`?

Reply:
Going through `.child()` ensures that an IndexError is raised if you go too far.

Reply:
Got it

Reply:
Wait, it's impossible to go to far for `before()` or `after()`...

Reply:
Oh, I see if dist is not 1, nm.